### PR TITLE
add support for LOB variables and file reference variables

### PR DESCRIPTION
**Pull request by ilidio-lopes**
## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [x] Check the test cases in the PR . Host variables can be in a sample code

```cobol
        Identification Division.
        Program-Id. 'TEST1'.
        Data Division.
         Working-Storage Section.
       01  VAR-NAME SQL TYPE IS BINARY LARGE OBJECT (30).
       01  VAR-NAME USAGE IS SQL TYPE IS BLOB-LOCATOR.
       01  VAR-NAME USAGE IS SQL TYPE IS BINARY LARGE OBJECT (30 M).
        PROCEDURE division
```
```cobol
        Identification Division.
        Program-Id. 'TEST1'.
        Data Division.
        Working-Storage Section.
       01 GREET. 
           04 VAR1 USAGE IS SQL TYPE IS CHARACTER LARGE OBJECT (10 M).
        PROCEDURE division.
           display  VAR1.
```

image

- [x] autocomplete
image

- [ ] Test A
- [ ] Test B

## Checklist:
- [x] Each of my commits contains one meaningful change
- [x] I have performed rebase of my branch on top of the development
- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] I have checked my code and corrected any misspellings

### CODE REVIEW

File: server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/sql/TestSqlHostVariable.java [hunk 1]
```diff
@@ -52,6 +52,26 @@ public class TestSqlHostVariable {
           + "        PROCEDURE DIVISION.\n"
           + "         {_01 VAR-NAME USAGE IS SQL TYPE IS RESULT-SET-LOCATOR VARYING|1_}.";
 
+  public static final String TEXT4 =
+          "        Identification Division.\n"
+                  + "        Program-Id. 'TEST1'.\n"
+                  + "        Data Division.\n"
+                  + "         Working-Storage Section.\n"
+                  + "       01 {$*VAR-NAME1} USAGE IS SQL TYPE IS BINARY LARGE OBJECT (10).\n"
```
Comment (mm-broadcom) [inline server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/sql/TestSqlHostVariable.java]: Cannot be 01 for the level. The IBM documentation, confusingly, states that `level-1` in their grammar is for COBOL levels in the range 2-48: https://www.ibm.com/docs/en/db2-for-zos/13?topic=cobol-host-variable-arrays-in#db2z_hostvariablearraycobol__title__16

Comment (ilidio-lopes) [reply server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/sql/TestSqlHostVariable.java]: That is not the correct reference you sent. This is https://www.ibm.com/docs/en/db2-for-zos/13?topic=cobol-host-variables-in#db2z_hostvariablecobol__title__23
It can be 01 for level.

Comment (amanPras) [reply server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/sql/TestSqlHostVariable.java]: I would use the pdf attached in the issue https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/issues/2320

**Comment by amanPras** [comment]
I tried to compile below code on mainframe
```cobol
        Identification Division.
        Program-Id. 'TEST1'.
        Data Division.
         Working-Storage Section.
         LINKAGE SECTION.
      *  01 BIN-VAR USAGE IS SQL TYPE IS BINARY(10).
      *  01 VBIN-VAR USAGE IS SQL TYPE IS BINARY VARYING(10).
      *  01 VBIN-RES USAGE IS SQL TYPE IS RESULT-SET-LOCATOR VARYING.
        01 VAR-NAME1 USAGE IS SQL TYPE IS BINARY LARGE OBJECT (10).
        PROCEDURE division.
      *    display  VBIN-VAR-LEN.
      *    display  VBIN-VAR-TEXT.
      *    DISPLAY BIN-VAR.
``` 

seems this introduce/inject below var struture

```
      9   1  VAR-NAME1 . . . . . . . . . . . . . . . . . . BLL=00001   000000000   DS 0CL14          Group
      0     2  VAR-NAME1-LENGTH. . . . . . . . . . . . . . BLL=00001   000000000   DS 4C             Binary
      0     2  VAR-NAME1-DATA. . . . . . . . . . . . . . . BLL=00001   000000004   DS 10C            Display
```


 We can refer Db2ImplicitVariablesGenerator for a similar use case.

File: server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/sql/TestSqlHostVariable.java [hunk 2]
```diff
@@ -84,6 +84,28 @@ public class TestSqlHostVariable {
...
+                  + "       01 {$*VAR-NAME4} USAGE IS SQL TYPE IS CHAR LARGE OBJECT (10 G).\n"
+                  + "       01 {$*VAR-NAME5} USAGE IS SQL TYPE IS CLOB (20).\n"
+                  + "       01 {$*VAR-NAME6} USAGE IS SQL TYPE IS DBCLOB (30 K).\n"
+                  + "       01 {$*VAR-NAME7} USAGE IS SQL TYPE IS BLOB-LOCATOR.\n"
+                  + "       01 {$*VAR-NAME8} USAGE IS SQL TYPE IS CLOB-LOCATOR.\n"
+                  + "       01 {$*VAR-NAME9} USAGE IS SQL TYPE IS DBCLOB-LOCATOR.\n"
+                  + "       01 {$*VAR-NAME10} USAGE IS SQL TYPE IS BLOB-FILE.\n"
+                  + "       01 {$*VAR-NAME11} USAGE IS SQL TYPE IS CLOB-FILE.\n"
+                  + "       01 {$*VAR-NAME12} USAGE IS SQL TYPE IS DBCLOB-FILE.\n"
+                  + "       01 {$*VAR-NAME12} USAGE IS SQL TYPE IS BINARY LARGE OBJECT (10).\n"
```
Comment (amanPras) [inline server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/sql/TestSqlHostVariable.java]: @shciubrcom introduced a mechanism to check the definition location of these implicit variables.
Please check https://github.com/eclipse-che4z/che-che4z-lsp-for-cobol/blob/development/server/engine/src/test/java/org/eclipse/lsp/cobol/usecases/sql/TestSqlHostVariableVarbinary.java

**Comment by amanPras** [comment]
LGTM! except for the level issue we discussed

`1 level-1 indicates a COBOL level in the range 2–48.`

which results in the error for below compliable code
```cobol
        Identification Division.
        Program-Id. 'TEST1'.
        Data Division.
         Working-Storage Section.
         01 TST.
           05 TST-1.
                   10 TEST-11 pic x.
                   10 TEST-22 .
                           48 VBIN-VAR USAGE IS SQL TYPE IS 
                           BINARY VARYING(10).
        PROCEDURE division.
           display  VBIN-VAR-LEN.
           display  VBIN-VAR
           .
```

**Comment by amanPras** [comment]
Also, please resolve merge conflict

File: server/engine/src/main/antlr4/org/eclipse/lsp/db2/parser/Db2SqlParser.g4 [hunk 3]
```diff
@@ -38,6 +36,18 @@ binary_host_variable_varbinary_size: T=dbs_integerliteral_expanded {validateInte
 
 host_variable_usage: (USAGE IS?)? SQL TYPE IS;
 
+sql_host_variables: dbs_level_01 entry_name host_variable_usage (result_set_locator | lobWithSize | lobNoSize);
+
+result_set_locator : RESULT_SET_LOCATOR VARYING;
+
+lobWithSize
+    : (BINARY LARGE OBJECT | BLOB | CHARACTER LARGE OBJECT | CHAR LARGE OBJECT | CLOB | DBCLOB) LPARENCHAR dbs_integer k_m_g? RPARENCHAR
```
Comment (amanPras) [inline server/engine/src/main/antlr4/org/eclipse/lsp/db2/parser/Db2SqlParser.g4]: @ishche (question/ doubt) would it help if "BINARY LARGE OBJECT" is combined as one rule , from performance perspective ?

**Review by amanPras — APPROVED** [review]
LGTM

File: server/engine/src/test/java/org/eclipse/lsp/cobol/lsp/SourceUnitGraphTest.java [hunk 4]
```diff
@@ -103,10 +103,10 @@ void testLinksBetweenCopybookAndSourceCode() {
...
 
     assertEquals("COPY 3 TEXT", sourceUnitGraph.getCopyNodeContent(copyNode3));
-    assertTrue(sourceUnitGraph.isCopybook(copy1Uri));
-    assertTrue(sourceUnitGraph.isCopybook(copy2Uri));
-    assertTrue(sourceUnitGraph.isCopybook(copy3Uri));
-    assertFalse(sourceUnitGraph.isCopybook(URI));
+    assertTrue(sourceUnitGraph.isUserSuppliedCopybook(copy1Uri));
+    assertTrue(sourceUnitGraph.isUserSuppliedCopybook(copy2Uri));
+    assertTrue(sourceUnitGraph.isUserSuppliedCopybook(copy3Uri));
+    assertFalse(sourceUnitGraph.isUserSuppliedCopybook(URI));
```
Comment (amanPras) [inline server/engine/src/test/java/org/eclipse/lsp/cobol/lsp/SourceUnitGraphTest.java]: unrelated changes from 8bd8a7135735dd185a3a0a6aebd042b6e19260fb. Please rebase with latest development

**Review by amanPras — CHANGES_REQUESTED** [review]
unrelated changes from 8bd8a7135735dd185a3a0a6aebd042b6e19260fb

File: server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlVisitor.java [hunk 5]
```diff
@@ -106,7 +118,11 @@ private List<Node> createHostVariableDefinitionNode(ParserRuleContext ctx, Parse
                         getLocality(this.context.getExtendedDocument().mapLocation(constructRange(nameCtx)))))
                 .usageClauses(ImmutableList.of(UsageFormat.DISPLAY))
                 .build();
-
+        if (ctx instanceof  Db2SqlParser.Lob_host_variablesContext
+                && ((Db2SqlParser.Lob_host_variablesContext) ctx).lobWithSize() != null
+                && ((Db2SqlParser.Lob_host_variablesContext) ctx).lobWithSize().BINARY() != null) {
+            generateVarbinVariables(variableDefinitionNode, ((Db2SqlParser.Lob_host_variablesContext) ctx).lobWithSize().dbs_integer().getText(), ctx);
```
Comment (amanPras) [inline server/engine/src/main/java/org/eclipse/lsp/cobol/implicitDialects/sql/Db2SqlVisitor.java]: Just a thought, this puts extra responsibility to `createHostVariableDefinitionNode` . Should this go to the actual visit method that calls this method ?

**Review by amanPras — APPROVED** [review]

**Comment by amanPras** [comment]
#2320
